### PR TITLE
Add options and zone possibilities

### PR DESCRIPTION
**Pull request by AblionGE**
I added diverse options for bind that can be necessary to manage your own DNS :
- recursion
- allow_query
- dnssec_enable
- dnssec_validation

And also the possibility to not create a zone file when you declare a zone (only add an entry in the conf file)

An finally, I adapted the templates.

### CODE REVIEW

**Comment by ekohl** [comment]
Looks like it needs a rebase.

File: manifests/init.pp [hunk 1]
```diff
@@ -12,9 +12,13 @@
...
   $forwarders           = $::dns::params::forwarders,
   $listen_on_v6         = $::dns::params::listen_on_v6,
+  $recursion            = $::dns::params::recursion,
   $allow_recursion      = $::dns::params::allow_recursion,
+  $allow_query          = $::dns::params::allow_query,
+  $dnssec_enable        = $::dns::params::dnssec_enable,
+  $dnssec_validation    = $::dns::params::dnssec_validation,
   $namedconf_template   = $::dns::params::namedconf_template,
-  $optionsconf_template = $::dns::params::optionsconf_template,
+  $optionsconf_template = $::dns::params::optionsconf_template
```
Comment (mmoll) [inline manifests/init.pp]: the comma can stay, as it eases future extensions ;)

**Comment by mmoll** [comment]
I didn't test this, but it's looking correct code-wise, just one nitpick.
Maybe it would make sense to use booleans for the yes/no values?

**Comment by AblionGE** [comment]
For the yes/no values, my idea was to be close of the actual values in the bind documentation.

For example : http://www.zytrax.com/books/dns/ch7/queries.html#recursion

I wanted to write tests, but I don't know when I will have time fot that...

**Comment by mmoll** [comment]
For the yes/no values, my idea is to use the boolean type in puppet because of validation and the whole handling regarding type safety (if one may call it like that) in Puppet.

Regarding tests, this module has an embarrassingly low coverage, so I guess we can merge this without spec tests for now and cry later :cry:.

**Comment by mmoll** [comment]
Oh, and please squash your commits together into one.

**Comment by AblionGE** [comment]
Squash done!
About the validation of yes/no values, I added some validations of regular expressions in the init class to enforce it. This remains a string but I think it's better to stay close of the actual values of bind.

File: manifests/init.pp [hunk 2]
```diff
@@ -12,12 +12,20 @@
...
+  $allow_query          = $::dns::params::allow_query,
+  $dnssec_enable        = $::dns::params::dnssec_enable,
+  $dnssec_validation    = $::dns::params::dnssec_validation,
   $namedconf_template   = $::dns::params::namedconf_template,
   $optionsconf_template = $::dns::params::optionsconf_template,
 ) inherits dns::params {
   validate_array($dns::forwarders)
   validate_array($dns::allow_recursion)
+  validate_array($dns::allow_query)
+  validate_re($dns::recursion, ['yes', 'no'], 'Only \'yes\' and \'no\' are valide values for recursion field')
```
Comment (domcleal) [inline manifests/init.pp]: a) use ^ and $ in regexes to anchor it, e.g. `'^yes$'` or even combine them with '^(yes|no)$'.
b) "valide" should be "valid"

**Comment by domcleal** [comment]
I don't really have an opinion on bool/string I'm afraid.  I would usually prefer boolean, but it wouldn't surprise me if BIND added a third value to those fields in future, so string allows for a bit more flexibility.

We don't seem to have any precedent in this module that I can see to help..

File: manifests/init.pp (hunk 2 above)
Comment (AblionGE) [inline manifests/init.pp]: Yes of course ;)

**Comment by AblionGE** [comment]
Exactly, that was what I had in mind. Future extension of BIND will be easier to include

File: manifests/zone.pp [hunk 3]
```diff
@@ -16,10 +16,14 @@
...
+    $file           = true,
+    $forward        = 'first',
+    $forwarders     = [],
 ) {
 
-  validate_bool($reverse)
-  validate_array($masters, $allow_transfer)
+  validate_bool($reverse, $file)
+  validate_array($masters, $allow_transfer, $forwarders)
+  validate_re($forward, ['first', 'only'], 'Only \'first\' or \'only\' are valid values for forward field')
```
Comment (mmoll) [inline manifests/zone.pp]: '^(first|only)$' then here also, please

**Comment by mmoll** [comment]
one last validate_re, then I'm :+1:

**Comment by AblionGE** [comment]
Done :+1:

File: manifests/zone.pp [hunk 4]
```diff
@@ -16,10 +16,14 @@
     $contact        = "root.${title}.",
     $zonefilepath   = $::dns::zonefilepath,
     $filename       = "db.${title}",
+    $file           = true,
```
Comment (ekohl) [inline manifests/zone.pp]: This is somewhat generic. Maybe name it ```manage_file``` or something similar?

File: manifests/params.pp [hunk 5]
```diff
@@ -57,5 +57,10 @@
 
     $listen_on_v6         = 'any'
 
+    $recursion            = 'yes'
     $allow_recursion      = []
+    $allow_query          = [ 'any' ]
+
+    $dnssec_enable        = 'no'
+    $dnssec_validation    = 'no'
```
Comment (ekohl) [inline manifests/params.pp]: http://ftp.isc.org/isc/bind9/cur/9.8/doc/arm/Bv9ARM.ch06.html states the default is ```yes``` and I'd prefer to match defaults.

**Comment by ekohl** [comment]
For the record, I agree with yes/no instead of booleans.

File: manifests/params.pp (hunk 5 above)
Comment (AblionGE) [inline manifests/params.pp]: It seems obvious ;)

File: manifests/zone.pp (hunk 4 above)
Comment (AblionGE) [inline manifests/zone.pp]: :+1:

**Comment by ekohl** [comment]
:+1:

**Comment by mmoll** [comment]
merged, merci @AblionGE!

**Pull request merged**